parallel_cram.sh

```diff
@@ -0,0 +1,185 @@
+            fi
+            shift
+            ;;
+        -j*)  # Handle -j4 format
+            PARALLEL_JOBS="${1#-j}"
+            if ! [[ "$PARALLEL_JOBS" =~ ^[0-9]+$ ]]; then
+                echo "Error: -j requires a numeric argument"
+                exit 1
+            fi
+            if [ "$PARALLEL_JOBS" -lt 1 ]; then
+                echo "Error: Number of jobs must be at least 1"
+                exit 1
+            fi
+            shift
+            ;;
+        --keep-tmpdir)
+            KEEP_TMPDIR=1
+            shift
+            ;;
+        --shell=*)
+            SHELL_PATH="${1#*=}"
+            shift
+            ;;
+        --shell-opts=*)
+            SHELL_OPTS="${1#*=}"
+            shift
+            ;;
+        --indent=*)
+            INDENT="${1#*=}"
+            shift
+            ;;
+        -*)
+            echo "Unknown option: $1"
+            usage
+            ;;
+        *)
+            TEST_DIR="$1"
+            shift
+            ;;
+    esac
+done
+
+# Validate test directory
+if [ ! -d "$TEST_DIR" ]; then
+    echo "Error: Directory $TEST_DIR does not exist"
+    exit 1
+fi
+
+# Build cram command with options
+CRAM_CMD="cram"
+[ $QUIET -eq 1 ] && CRAM_CMD="$CRAM_CMD -q"
+[ $VERBOSE -eq 1 ] && CRAM_CMD="$CRAM_CMD -v"
+[ $KEEP_TMPDIR -eq 1 ] && CRAM_CMD="$CRAM_CMD --keep-tmpdir"
+[ -n "$SHELL_PATH" ] && CRAM_CMD="$CRAM_CMD --shell=$SHELL_PATH"
+[ -n "$SHELL_OPTS" ] && CRAM_CMD="$CRAM_CMD --shell-opts=$SHELL_OPTS"
+[ -n "$INDENT" ] && CRAM_CMD="$CRAM_CMD --indent=$INDENT"
+
+# Find all .t files and order with priority files first
+TEST_FILES=$(find "$TEST_DIR" -name "*.t" | sort)
+PRIORITY_FILES=$(echo "$TEST_FILES" | grep -E "$PRIORITY_PATTERNS" || true)
+OTHER_FILES=$(echo "$TEST_FILES" | grep -vE "$PRIORITY_PATTERNS" || true)
+ALL_FILES=$(echo -e "${PRIORITY_FILES}\n${OTHER_FILES}")
+
+# Create a lock directory for synchronized output
+LOCK_DIR="/tmp/cram_lock_$$"
+mkdir "$LOCK_DIR"
+trap 'rm -rf "$LOCK_DIR"' EXIT
+
+# Create files to store temporary directories and test results
+TMPDIR_FILE="$LOCK_DIR/tmpdirs"
+RESULTS_FILE="$LOCK_DIR/results"
+touch "$TMPDIR_FILE" "$RESULTS_FILE"
+
+# Export variables for subshell
+export CRAM_CMD LOCK_DIR TMPDIR_FILE RESULTS_FILE KEEP_TMPDIR
+
+# Run tests in parallel with specified number of jobs
+echo "$ALL_FILES" | xargs -P "$PARALLEL_JOBS" -I {} sh -c '
+    file="$1"
+    lock_dir="$LOCK_DIR"
+    output=$($CRAM_CMD "$file" 2>&1)
+    status=$?
+    
+    while ! mkdir "$lock_dir/lock" 2>/dev/null; do
+        sleep 0.05
+    done
+    
+    # Record test status
+    echo "$status" >> "$RESULTS_FILE"
+    
+    if [ $status -eq 0 ]; then
+        printf "."
+    else
+        echo "$output" | grep -v "^# Ran "
+    fi
+    
+    # If --keep-tmpdir is enabled, capture the temporary directory
+    if [ $KEEP_TMPDIR -eq 1 ]; then
+        echo "$output" | grep "Kept temporary directory:" >> "$TMPDIR_FILE"
+    fi
+    
+    rm -rf "$lock_dir/lock"
+' - {}
```

Comment:
FWIW, this reimplements a bit of functionality in `parallel` (which is typically more useful than `xargs`).